File: src/main/java/com/tang/intellij/lua/stubs/index/LuaClassMethodIndex.java
/*
 * Copyright (c) 2017. tangzx(love.tangzx@qq.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.tang.intellij.lua.stubs.index;

import com.intellij.psi.stubs.StringStubIndexExtension;
import com.intellij.psi.stubs.StubIndexKey;
import com.tang.intellij.lua.lang.LuaLanguage;
import com.tang.intellij.lua.psi.LuaClassMethodDef;
import com.tang.intellij.lua.search.SearchContext;
import org.jetbrains.annotations.NotNull;

import java.util.Collection;

/**
 *
 * Created by tangzx on 2016/12/4.
 */
public class LuaClassMethodIndex extends StringStubIndexExtension<LuaClassMethodDef> {
    public static final StubIndexKey<String, LuaClassMethodDef> KEY = StubIndexKey.createIndexKey("lua.index.class.method");

    private static final LuaClassMethodIndex INSTANCE = new LuaClassMethodIndex();

    public static LuaClassMethodIndex getInstance() {
        return INSTANCE;
    }

    @Override
    public int getVersion() { return LuaLanguage.INDEX_VERSION;}

    @NotNull
    @Override
    public StubIndexKey<String, LuaClassMethodDef> getKey() {
        return KEY;
    }

    public static Collection<LuaClassMethodDef> findStaticMethods(@NotNull String className, @NotNull SearchContext context) {
        String key = className + ".static";
        return INSTANCE.get(key, context.getProject(), context.getScope());
    }

    public static LuaClassMethodDef findStaticMethod(@NotNull String className, @NotNull String methodName, @NotNull SearchContext context) {
        if (context.isDumb())
            return null;

        Collection<LuaClassMethodDef> collection = INSTANCE.get(className + ".static." + methodName, context.getProject(), context.getScope());
        if (collection.isEmpty())
            return null;
        else
            return collection.iterator().next();
    }

    public static LuaClassMethodDef findMethodWithName(@NotNull String className, @NotNull String methodName, @NotNull SearchContext context) {
        if (context.isDumb())
            return null;

        Collection<LuaClassMethodDef> collection = INSTANCE.get(className, context.getProject(), context.getScope());
        for (LuaClassMethodDef methodDef : collection) {
            if (methodName.equals(methodDef.getName())) {
                return methodDef;
            }
        }
        return null;
    }
}

File: src/main/java/com/tang/intellij/lua/stubs/index/LuaClassMethodIndex.kt
/*
 * Copyright (c) 2017. tangzx(love.tangzx@qq.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.tang.intellij.lua.stubs.index

import com.intellij.openapi.project.Project
import com.intellij.psi.search.GlobalSearchScope
import com.intellij.psi.stubs.StringStubIndexExtension
import com.intellij.psi.stubs.StubIndex
import com.intellij.psi.stubs.StubIndexKey
import com.intellij.util.Processor
import com.tang.intellij.lua.lang.LuaLanguage
import com.tang.intellij.lua.psi.LuaClassMethod
import com.tang.intellij.lua.search.SearchContext

/**
 *
 * Created by tangzx on 2016/12/4.
 */
class LuaClassMethodIndex : StringStubIndexExtension<LuaClassMethod>() {

    override fun getVersion(): Int {
        return LuaLanguage.INDEX_VERSION
    }

    override fun getKey(): StubIndexKey<String, LuaClassMethod> {
        return KEY
    }

    override fun get(s: String, project: Project, scope: GlobalSearchScope): Collection<LuaClassMethod> {
        return StubIndex.getElements(KEY, s, project, scope, LuaClassMethod::class.java)
    }

    companion object {
        val KEY = StubIndexKey.createIndexKey<String, LuaClassMethod>("lua.index.class.method")

        val instance = LuaClassMethodIndex()

        fun findStaticMethods(className: String, context: SearchContext): Collection<LuaClassMethod> {
            val key = className + ".static"
            return instance.get(key, context.project, context.getScope())
        }

        fun process(key: String, context: SearchContext, processor: Processor<LuaClassMethod>): Boolean {
            if (context.isDumb)
                return false
            return StubIndex.getInstance().processElements(KEY, key, context.project, context.getScope(), null, LuaClassMethod::class.java, processor)
        }

        fun findStaticMethod(className: String, methodName: String, context: SearchContext): LuaClassMethod? {
            if (context.isDumb)
                return null

            val key = className + ".static." + methodName
            var result: LuaClassMethod? = null
            process(key, context, Processor {
                result = it
                false
            })
            return result
        }

        fun findMethodWithName(className: String, methodName: String, context: SearchContext): LuaClassMethod? {
            if (context.isDumb)
                return null
            var result: LuaClassMethod? = null
            process(className, context, Processor {
                if (methodName == it.name) {
                    result = it
                    return@Processor false
                }
                true
            })
            return result
        }
    }
}